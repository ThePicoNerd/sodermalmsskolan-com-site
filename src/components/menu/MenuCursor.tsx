import { capitalize } from "lodash";
import { DateTime } from "luxon";
import React, { FunctionComponent } from "react";
import { MenuScope, useMenuContext } from "../../lib/food/MenuContext";
import InlineSkeleton from "../skeleton/InlineSkeleton";

export interface MenuCursorProps {
  cursor?: DateTime;
  scope?: MenuScope;
  showYear?: boolean;
}

/**
 * A cursor for the menu.
 *
 * @param {React.PropsWithChildren<MenuCursorProps>} props Props.
 *
 * @returns {React.ReactElement} The rendered cursor.
 */
const MenuCursor: FunctionComponent<MenuCursorProps> = ({
  cursor: forcedCursor,
  scope: forcedScope,
  showYear = false,
}) => {
  const { cursor: contextCursor, scope: contextScope } = useMenuContext();

  const cursor = forcedCursor ?? contextCursor;
  const scope = forcedScope ?? contextScope;

  if (!cursor) {
    return (
      <InlineSkeleton width="6em" />
    );
  }

  switch (scope) {
    case "week":
      return (
        <>
          Vecka
          {" "}
          {cursor?.weekNumber}
          {showYear ? ` ${cursor?.year}` : undefined}
        </>
      );
    case "month":
      return (
        <>
          {capitalize(cursor.toLocaleString({
            month: "long",
            year: "numeric",
          }))}
        </>
      );
    default:
      return (
        <>
          {cursor?.toLocaleString(DateTime.DATE_FULL)}
        </>
      );
  }
};

export default MenuCursor;
