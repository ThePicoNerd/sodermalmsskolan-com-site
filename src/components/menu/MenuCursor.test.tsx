import { render } from "@testing-library/react";
import { capitalize } from "lodash";
import { DateTime } from "luxon";
import React from "react";
import MenuCursor from "./MenuCursor";

describe("<MenuCursor /> tests", () => {
  it("should work as expected", () => {
    const date = DateTime.utc(2020, 1, 1);

    const weekResult = render(<MenuCursor cursor={date} scope="week" />);
    expect(weekResult.asFragment().textContent).toBe("Vecka 1");

    const weekWithYearResult = render(<MenuCursor cursor={date} scope="week" showYear />);
    expect(weekWithYearResult.asFragment().textContent).toBe("Vecka 1 2020");

    const monthResult = render(<MenuCursor cursor={date} scope="month" />);
    expect(monthResult.asFragment().textContent).toBe(capitalize(date.toLocaleString({
      month: "long",
      year: "numeric",
    })));

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const fallbackResult = render(<MenuCursor cursor={date} scope={"invalid" as any} />);
    expect(fallbackResult.asFragment().textContent).toBe(date.toLocaleString(DateTime.DATE_FULL));
  });

  it("should fall back to skeletons", () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const { asFragment } = render(<MenuCursor cursor={false as any} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
