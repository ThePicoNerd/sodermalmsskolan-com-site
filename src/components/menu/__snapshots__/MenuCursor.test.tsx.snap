// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuCursor /> tests should fall back to skeletons 1`] = `
<DocumentFragment>
  <span>
    <span
      class="skeleton max-w-full"
      style="width: 6em;"
    >
      â€Œ
    </span>
  </span>
</DocumentFragment>
`;
